app/common/src/utilities/data/reactive.ts

```diff
@@ -0,0 +1,104 @@
+import { isRef, unref, type ComputedRef, type Ref } from 'vue'
+
+/**
+ * A value that may be wrapped in a ref. Note that this is different from the Vue type with the same
+ * name.
+ */
+type MaybeRef<T> = Ref<T> | ComputedRef<T> | T
+
+/**
+ * A type that may recursively contain {@link Ref}s or getter functions, and can be recursively
+ * unwrapped by {@link cloneDeepUnref}
+ */
+export type MaybeRefDeep<T> = MaybeRef<
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
+  T extends Function ? T
+  : T extends object ?
+    {
+      [Property in keyof T]: MaybeRefDeep<T[Property]>
+    }
+  : T
+>
+
+/**
+ * Helper function for cloning deep objects where the level and key is provided to the callback
+ * function.
+ */
+function _cloneDeep<T>(
+  value: MaybeRefDeep<T>,
+  customize?: (val: MaybeRefDeep<T>, key: string, level: number) => T | undefined,
+  currentKey: string = '',
+  currentLevel: number = 0,
+): T {
+  if (customize) {
+    const result = customize(value, currentKey, currentLevel)
+    if (result === undefined && isRef(value)) {
+      return result as T
+    }
+    if (result !== undefined) {
+      return result
+    }
+  }
+
+  if (Array.isArray(value)) {
+    return value.map((val, index) =>
+      _cloneDeep(val, customize, String(index), currentLevel + 1),
+    ) as unknown as T
+  }
+
+  if (typeof value === 'object' && isPlainObject(value)) {
+    const entries = Object.entries(value).map(([key, val]) => [
+      key,
+      _cloneDeep(val, customize, key, currentLevel + 1),
+    ])
+    return Object.fromEntries(entries)
+  }
+
+  return value as T
+}
+
+function cloneDeep<T>(
+  value: MaybeRefDeep<T>,
+  customize?: (val: MaybeRefDeep<T>, key: string, level: number) => T | undefined,
+): T {
+  return _cloneDeep(value, customize)
+}
```

Comment:
I don't know why we have separate `cloneDeep` and `_cloneDeep` since the former is just a wrapper for the latter disallowing some arguments, and both are private anyway.